components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java

```diff
@@ -181,34 +171,67 @@ protected void doStop() throws Exception {
         if (iterator != null) {
             final Iterator<Object> msgList = iterator;
             final String msgTopic = topic;
-            return new Iterator<ProducerRecord>() {
+            return new Iterator<Pair<Object, ProducerRecord>>() {
                 @Override
                 public boolean hasNext() {
                     return msgList.hasNext();
                 }
 
                 @Override
-                public ProducerRecord next() {
+                public Pair<Object, ProducerRecord> next() {
```

Comment:
Sorry, I am bit confused here, was there a logical reason to use `pair` with the message body?

Reply:
If the left of pair is instance of Message or Exchange it will then be used to set metadata on this Exchange/Message instance. If this is instance of something else, no metadata will be set at all.

If we return only the ProducerRecord, we loose information to what Message/Exchange it refers to and we cannot set metadata returned by Kafka producer.